✨ Base stuff (models, tests and such)

merca opened the pull request:
All new reviewers: This repo uses bunch of automated python reviews as well so the amount of comments can be really high - specially in the beginning when I just start out as rookie 😆 


## In this PR
- [x] add data models

## TODO list

- [ ] add login
- [ ] add find Aduro stove by manufacturer
- [ ] add get device info
- [ ] add get device entities
- [ ] add get entity states
- [ ] add update entity state if allowed

Commit:
✨ Add data models and tests

Commit:
♻ recfactor and resolve 🐶 problems

Commit:
🐶 fix rest of the review dog problems

Commit:
🔰 make pylint happy

Commit:
why is pytest not agreeng with module scoped fixtures

Commit:
and fix isort

Commit:
BInary else is supported in 3.10

Commit:
add test for helper method

Commit:
use `is` instead `==` on singleton value

Commit:
Tell Sonar to use python 3

Commit:
Fix unused import

Github testing does not count for module scoped fictures

Commit:
use __version__ in representing the class

Commit:
Use verbal pylint exclude

and add code coverage to codacy

Commit:
Simplify assert for empty list

After changing the List declaration in model, using `assert not member` is preferred style

Commit:
Fix failing auto create issues

theovald commented on aduro/helpers.py:
```diff
@@ -0,0 +1,25 @@
+                    return True
+                if orig_value == "false":
+                    return False
+                raise ValueError from err
```
This really deep nesting of ValueErrors makes me a little dizzy. Possible to de-nest in some way?

merca replied:
I was thinking aboutt it my self and I'm glad you made me change it 😃

theovald reviewed (commented):
Great stuff 👍 

All of the from_dict unpacking has me wondering if this may be a really good use case for employing [pydantic](https://pydantic-docs.helpmanual.io/). The parsing would probably require less code duplication (`str(obj.get("field")`), and you get type hint *enforcement*, not just hint.

The structure seams very familiar and clean 👏

Commit:
reduce nesting in `try_convert_object`

- remove nesting, use if/else
- add some more tests to make sure also `None`will be covered

merca commented:
Thanx @theovald for the tip - I'm gonna try this. I try to keep custom packages to  minimal, since this will be used in Home Assistant, but it really can reduce som code here 😛.


> Great stuff 👍
> 
> All of the from_dict unpacking has me wondering if this may be a really good use case for employing [pydantic](https://pydantic-docs.helpmanual.io/). The parsing would probably require less code duplication (`str(obj.get("field")`), and you get type hint _enforcement_, not just hint.
> 
> The structure seams very familiar and clean 👏

Tenrec commented on aduro/model.py:
```diff
@@ -0,0 +1,239 @@
+    online: bool
+
+    @staticmethod
+    def from_dict(obj: Any) -> "Connection":
```
Maybe I'm not 100% sure on the syntax, but shouldn't it be like this?

```suggestion
    def from_dict(obj: Any) -> Connection:
```

merca replied:
@Tenrec :: Now I why I should have left comment on this when I had to change it, because now I don't remember anymore why it had to be in quotes 😆 
I'm changing this to `pydantic.dataclasses` to let this be handled by module.

Tenrec replied:
Then I'll hold back on any more comments on the dataclasses until after this rewrite. :)

merca replied:
@Tenrec now the dataclasses are converted to `pydantic`. It does look better, but I have not found yet how to convert the `data` field in `State` to common, since it always comes as string from api. Have to dig into theyr code to find out how to use custom converter. In worse case I use custom type and convert that on init with my converter 😄

Tenrec commented on aduro/model.py:
```diff
@@ -0,0 +1,239 @@
+    online: bool
+
+    @staticmethod
+    def from_dict(obj: Any) -> "Connection":
```
This makes your intent clearer. 

```suggestion
    def from_dict(obj: dict) -> "Connection":
```

Tenrec replied:
And a curse on the house of Python for its lack of multiple constructors. :)

Tenrec commented on aduro/model.py:
```diff
@@ -0,0 +1,239 @@
+# pylint: disable=too-many-instance-attributes
+@dataclass
+class Entity:
+    """Entity class for device entitie"""
```
```suggestion
    """Entity class for device entity"""
```

Commit:
switch to `pydantic`

To reduce some of the custom logic, I'm using `pydantic` to let it to takke care of the mapping

Tenrec commented on aduro/helpers.py:
```diff
@@ -0,0 +1,23 @@
+        value = int(orig_value)
+    else:
+        value = orig_value
+    return value
```
Should this return a value that we don't know what is, or instead throw an exception?

merca replied:
I would like it to return value anyhow, but maybe convert this to str in this case?

Tenrec replied:
That seems safer, and then you can also use Union[int, float, bool, str, ...] for the return type.

Commit:
🔰 Pylint does not like me again

merca commented:
> Great stuff 👍
> 
> All of the from_dict unpacking has me wondering if this may be a really good use case for employing [pydantic](https://pydantic-docs.helpmanual.io/). The parsing would probably require less code duplication (`str(obj.get("field")`), and you get type hint _enforcement_, not just hint.
> 
> The structure seams very familiar and clean 👏

@theovald Now I have converted to `pydantic` and it does look better, only issue I have is that the `data` object in `State` class is given as string and I would like to convert this to int/float/bool if content suggest it. But have not found how to plug in custom converter yet - maybe you know this?

Commit:
Typed return for `helpers.try_convert_object`

theovald commented:
> > Great stuff 👍
> > All of the from_dict unpacking has me wondering if this may be a really good use case for employing [pydantic](https://pydantic-docs.helpmanual.io/). The parsing would probably require less code duplication (`str(obj.get("field")`), and you get type hint _enforcement_, not just hint.
> > The structure seams very familiar and clean 👏
> 
> @theovald Now I have converted to `pydantic` and it does look better, only issue I have is that the `data` object in `State` class is given as string and I would like to convert this to int/float/bool if content suggest it. But have not found how to plug in custom converter yet - maybe you know this?

Ah, no, actually, have not run into this the few times I was lucky enough to work with pydantic. Only with clean, consistent data and datatypes. I see you added some returns on the try_convert method, was this related?

merca commented:
> > > Great stuff 👍
> > > All of the from_dict unpacking has me wondering if this may be a really good use case for employing [pydantic](https://pydantic-docs.helpmanual.io/). The parsing would probably require less code duplication (`str(obj.get("field")`), and you get type hint _enforcement_, not just hint.
> > > The structure seams very familiar and clean 👏
> > 
> > 
> > @theovald Now I have converted to `pydantic` and it does look better, only issue I have is that the `data` object in `State` class is given as string and I would like to convert this to int/float/bool if content suggest it. But have not found how to plug in custom converter yet - maybe you know this?
> 
> Ah, no, actually, have not run into this the few times I was lucky enough to work with pydantic. Only with clean, consistent data and datatypes. I see you added some returns on the try_convert method, was this related?

@theovald Lucky you!
`try_convert` will hopefully be used for this conversion in the end yes, when I find out how to plug this in. As for now - I will accept that all values will be defined as strings. Luckily for me -  the `entity` object will have data-type defined, so the end consumer may have to take the string and convert if I cant do this in `pydantric`. Will keep the method until confirmed I can't plug custom converter.
